From e0739ec1b5880d76d45be928eaa5dfd5dd64ccf5 Mon Sep 17 00:00:00 2001
From: Rebecca Chang Swee Fun <rebecca.swee.fun.chang@intel.com>
Date: Tue, 21 Jun 2016 14:11:21 +0800
Subject: [PATCH] configure.ac: add GTK_DOC_CHECK

Signed-off-by: Rebecca Chang Swee Fun <rebecca.swee.fun.chang@intel.com>
---
 configure.ac | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/configure.ac b/configure.ac
index 829f402..5509bfa 100644
--- a/configure.ac
+++ b/configure.ac
@@ -53,6 +53,8 @@ AM_CONDITIONAL(HAVE_GMODULE, test "x$have_gmodules" != "xno")
 PKG_CHECK_MODULES(GTK_X11, [gtk+-x11-3.0 x11], [have_gtk_x11="yes"], [have_gtk_x11="no"])
 AM_CONDITIONAL(HAVE_GTK_X11, test "x$have_gtk_x11" != "xno")
 
+GTK_DOC_CHECK([1.12])
+
 WARNING_CFLAGS="-Wall"
 
 AC_ARG_ENABLE(more-warnings,
-- 
1.9.1

