From 9b2dc96b9909e73f64fd0cb1d4612e767e7bf6bd Mon Sep 17 00:00:00 2001
From: Fuwei Tang <fuweix.tang@intel.com>
Date: Thu, 5 Sep 2019 09:05:10 +0800
Subject: [PATCH] msdkenc: work-around to avoid zero fps in MediaSDK structure

---
 sys/msdk/gstmsdkenc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/sys/msdk/gstmsdkenc.c b/sys/msdk/gstmsdkenc.c
index 08b462eb6..cf72d7472 100644
--- a/sys/msdk/gstmsdkenc.c
+++ b/sys/msdk/gstmsdkenc.c
@@ -316,6 +316,11 @@ gst_msdkenc_init_encoder (GstMsdkEnc * thiz)
         g_assert_not_reached ();
         break;
     }
+
+    /* work-around to avoid zero fps in msdk structure */
+    if (0 == thiz->vpp_param.vpp.In.FrameRateExtN)
+      thiz->vpp_param.vpp.In.FrameRateExtN = 30;
+
     thiz->vpp_param.vpp.Out = thiz->vpp_param.vpp.In;
     thiz->vpp_param.vpp.Out.FourCC = MFX_FOURCC_NV12;
     thiz->vpp_param.vpp.Out.ChromaFormat = MFX_CHROMAFORMAT_YUV420;
@@ -415,6 +420,10 @@ gst_msdkenc_init_encoder (GstMsdkEnc * thiz)
     thiz->param.mfx.FrameInfo.BitDepthChroma = 8;
   }
 
+  /* work-around to avoid zero fps in msdk structure */
+  if (0 == thiz->param.mfx.FrameInfo.FrameRateExtN)
+    thiz->param.mfx.FrameInfo.FrameRateExtN = 30;
+
   /* ensure bitrate control parameters */
   ensure_bitrate_control (thiz);
 
-- 
2.23.0

