From f8912747f3df628f248fac593362980fddcfb4f6 Mon Sep 17 00:00:00 2001
From: Chong Yi Chai <chong.yi.chai@intel.com>
Date: Wed, 4 Sep 2019 19:06:11 +0800
Subject: [PATCH] msdk plugin ranking is set to higher ranking

Signed-off-by: Jak Jailani, Nor Syahira <nor.syahira.jak.jailani@intel.com>
Signed-off-by: Chong Yi Chai <chong.yi.chai@intel.com>
---
 sys/msdk/gstmsdk.c | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/sys/msdk/gstmsdk.c b/sys/msdk/gstmsdk.c
index 1047ebb..52ae7b8 100644
--- a/sys/msdk/gstmsdk.c
+++ b/sys/msdk/gstmsdk.c
@@ -102,44 +102,44 @@ plugin_init (GstPlugin * plugin)
   if (!msdk_is_available ())
     return FALSE;
 
-  ret = gst_element_register (plugin, "msdkh264dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh264dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH264DEC);
 
-  ret = gst_element_register (plugin, "msdkh264enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh264enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH264ENC);
 
-  ret = gst_element_register (plugin, "msdkh265dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh265dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH265DEC);
 
-  ret = gst_element_register (plugin, "msdkh265enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh265enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH265ENC);
 
-  ret = gst_element_register (plugin, "msdkmjpegdec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmjpegdec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMJPEGDEC);
 
-  ret = gst_element_register (plugin, "msdkmjpegenc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmjpegenc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMJPEGENC);
 
-  ret = gst_element_register (plugin, "msdkmpeg2dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmpeg2dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMPEG2DEC);
 
-  ret = gst_element_register (plugin, "msdkmpeg2enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmpeg2enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMPEG2ENC);
 
-  ret = gst_element_register (plugin, "msdkvp8dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp8dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP8DEC);
 
-  ret = gst_element_register (plugin, "msdkvp8enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp8enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP8ENC);
 
-  ret = gst_element_register (plugin, "msdkvc1dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvc1dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVC1DEC);
 #ifdef USE_MSDK_VP9_DEC
-  ret = gst_element_register (plugin, "msdkvp9dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp9dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP9DEC);
 #endif
 
-  ret = gst_element_register (plugin, "msdkvpp", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvpp", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVPP);
 
   return ret;
-- 
2.22.0

