From 878a8c26c29f060c82d5bf568ad96407ed3593f8 Mon Sep 17 00:00:00 2001
From: Siddharth Ajit Waikar <siddharth.ajitx.waikar@intel.com>
Date: Thu, 5 Mar 2020 01:55:18 +0800
Subject: [PATCH] msdk plugin ranking is set to higher ranking

Signed-off-by: Siddharth Waikar <siddharth.ajitx.waikar@intel.com>
---
 sys/msdk/gstmsdk.c | 26 +++++++++++++-------------
 1 file changed, 13 insertions(+), 13 deletions(-)

diff --git a/sys/msdk/gstmsdk.c b/sys/msdk/gstmsdk.c
index fb92a6abb..cb253df87 100644
--- a/sys/msdk/gstmsdk.c
+++ b/sys/msdk/gstmsdk.c
@@ -104,44 +104,44 @@ plugin_init (GstPlugin * plugin)
   if (!msdk_is_available ())
     return FALSE;
 
-  ret = gst_element_register (plugin, "msdkh264dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh264dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH264DEC);
 
-  ret = gst_element_register (plugin, "msdkh264enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh264enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH264ENC);
 
-  ret = gst_element_register (plugin, "msdkh265dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh265dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH265DEC);
 
-  ret = gst_element_register (plugin, "msdkh265enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkh265enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKH265ENC);
 
-  ret = gst_element_register (plugin, "msdkmjpegdec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmjpegdec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMJPEGDEC);
 
-  ret = gst_element_register (plugin, "msdkmjpegenc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmjpegenc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMJPEGENC);
 
-  ret = gst_element_register (plugin, "msdkmpeg2dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmpeg2dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMPEG2DEC);
 
-  ret = gst_element_register (plugin, "msdkmpeg2enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkmpeg2enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKMPEG2ENC);
 
-  ret = gst_element_register (plugin, "msdkvp8dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp8dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP8DEC);
 
-  ret = gst_element_register (plugin, "msdkvc1dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvc1dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVC1DEC);
 #ifdef USE_MSDK_VP9_DEC
-  ret = gst_element_register (plugin, "msdkvp9dec", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp9dec", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP9DEC);
 #endif
 #ifdef USE_MSDK_VP9_ENC
-  ret = gst_element_register (plugin, "msdkvp9enc", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvp9enc", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVP9ENC);
 #endif
-  ret = gst_element_register (plugin, "msdkvpp", GST_RANK_NONE,
+  ret = gst_element_register (plugin, "msdkvpp", GST_RANK_PRIMARY + 3,
       GST_TYPE_MSDKVPP);
 
   return ret;
-- 
2.20.4

