diff -x .git -x .gitignore -Naur gstreamer-vaapi/gms-integration isg_gms-gstreamer-vaapi/gms-integration
--- gstreamer-vaapi/gms-integration	1970-01-01 07:30:00.000000000 +0730
+++ isg_gms-gstreamer-vaapi/gms-integration	2017-02-01 02:11:02.291633542 +0800
@@ -0,0 +1,5 @@
+###
+### GMS integration
+###
+upstream b60e2df63d8e1c214021dfb8a2bc302ce03eb97a
+	encoder: h264,h265: fix regression in offset count
