From 6fd8fb15b19d603575a56d2ffcae1302f1e301c6 Mon Sep 17 00:00:00 2001
From: "Ng, Mei Yeen" <mei.yeen.ng@intel.com>
Date: Tue, 4 Aug 2015 12:13:28 -0700
Subject: [PATCH] comment out malloc and realloc functions

Signed-off-by: Ng, Mei Yeen <mei.yeen.ng@intel.com>
---
 dsimple.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dsimple.c b/dsimple.c
index cd61870..f08828e 100644
--- a/dsimple.c
+++ b/dsimple.c
@@ -43,6 +43,7 @@ extern int screen;
 /*
  * Malloc: like malloc but handles out of memory using Fatal_Error.
  */
+/*
 char *Malloc(size)
      unsigned size;
 {
@@ -53,11 +54,12 @@ char *Malloc(size)
 
 	return(data);
 }
-	
+*/	
 
 /*
  * Realloc: like Malloc except for realloc, handles NULL using Malloc.
  */
+/*
 char *Realloc(ptr, size)
         char *ptr;
         int size;
@@ -72,7 +74,7 @@ char *Realloc(ptr, size)
 
 	return(new_ptr);
 }
-
+*/
 
 /*
  * Get_Display_Name (argc, argv) Look for -display, -d, or host:dpy (obselete)
-- 
1.9.1

