diff -x .git -x .gitignore -Naur xf86-video-intel/gms-integration isg_gms-sna_ddx/gms-integration
--- xf86-video-intel/gms-integration	1970-01-01 07:30:00.000000000 +0730
+++ isg_gms-sna_ddx/gms-integration	2017-02-01 02:10:28.155004248 +0800
@@ -0,0 +1,9 @@
+###
+### GMS integration
+###
+upstream 169c74fa6c2cd9c28dd7bfacd9639cd245b8c8a8
+	intel: Adding Marketing names for Skylake, Kabylake and Apollolake/Broxton.
+hsd/1504134673 9b4488f2453e2e3e8831906c2103631e6a807c95
+	Add NULL checking for drawable in sna_dri2_flip_event
+topic/vpg_stolen ed6285a9d39938572f40ba548625ff391ef0af9c
+	sna: Always clear entire gem create ioctl data
diff -x .git -x .gitignore -Naur xf86-video-intel/src/sna/kgem.c isg_gms-sna_ddx/src/sna/kgem.c
--- xf86-video-intel/src/sna/kgem.c	2017-02-01 02:10:47.803366438 +0800
+++ isg_gms-sna_ddx/src/sna/kgem.c	2017-02-01 02:10:28.179004690 +0800
@@ -879,8 +879,7 @@
 {
 	struct drm_i915_gem_create create;
 
-	VG_CLEAR(create);
-	create.handle = 0;
+	memset(&create, 0, sizeof(create));
 	create.size = PAGE_SIZE * num_pages;
 	(void)do_ioctl(fd, DRM_IOCTL_I915_GEM_CREATE, &create);
 
diff -x .git -x .gitignore -Naur xf86-video-intel/src/sna/sna_display.c isg_gms-sna_ddx/src/sna/sna_display.c
--- xf86-video-intel/src/sna/sna_display.c	2017-02-01 02:10:47.807366511 +0800
+++ isg_gms-sna_ddx/src/sna/sna_display.c	2017-02-01 02:10:28.183004764 +0800
@@ -697,8 +697,7 @@
 
 	assert((size & 4095) == 0);
 
-	VG_CLEAR(create);
-	create.handle = 0;
+	memset(&create, 0, sizeof(create));
 	create.size = size;
 	(void)drmIoctl(fd, DRM_IOCTL_I915_GEM_CREATE, &create);
 
diff -x .git -x .gitignore -Naur xf86-video-intel/src/sna/sna_dri2.c isg_gms-sna_ddx/src/sna/sna_dri2.c
--- xf86-video-intel/src/sna/sna_dri2.c	2017-02-01 02:10:47.807366511 +0800
+++ isg_gms-sna_ddx/src/sna/sna_dri2.c	2017-02-01 02:10:28.183004764 +0800
@@ -2948,6 +2948,10 @@
 	case FLIP_THROTTLE:
 		if (flip->signal) {
 			DBG(("%s: triple buffer swap complete, unblocking client\n", __FUNCTION__));
+			if(flip->draw == NULL) {
+				sna_dri2_event_free(flip);
+				break;
+			}
 			frame_swap_complete(flip, DRI2_FLIP_COMPLETE);
 		}
 	case FLIP_COMPLETE:
diff -x .git -x .gitignore -Naur xf86-video-intel/src/sna/sna_driver.c isg_gms-sna_ddx/src/sna/sna_driver.c
--- xf86-video-intel/src/sna/sna_driver.c	2017-02-01 02:10:47.807366511 +0800
+++ isg_gms-sna_ddx/src/sna/sna_driver.c	2017-02-01 02:10:28.187004838 +0800
@@ -406,8 +406,7 @@
 	if (res.count_crtcs == 0)
 		return TRUE;
 
-	VG_CLEAR(create);
-	create.handle = 0;
+	memset(&create, 0, sizeof(create));
 	create.size = 4096;
 	if (drmIoctl(fd, DRM_IOCTL_I915_GEM_CREATE, &create))
 		return FALSE;
diff -x .git -x .gitignore -Naur xf86-video-intel/test/dri3-test.c isg_gms-sna_ddx/test/dri3-test.c
--- xf86-video-intel/test/dri3-test.c	2017-02-01 02:10:47.819366733 +0800
+++ isg_gms-sna_ddx/test/dri3-test.c	2017-02-01 02:10:28.199005059 +0800
@@ -158,7 +158,7 @@
 {
 	struct drm_i915_gem_create create;
 
-	create.handle = 0;
+	memset(&create, 0, sizeof(create));
 	create.size = size;
 	(void)drmIoctl(fd, DRM_IOCTL_I915_GEM_CREATE, &create);
 
diff -x .git -x .gitignore -Naur xf86-video-intel/test/present-test.c isg_gms-sna_ddx/test/present-test.c
--- xf86-video-intel/test/present-test.c	2017-02-01 02:10:47.819366733 +0800
+++ isg_gms-sna_ddx/test/present-test.c	2017-02-01 02:10:28.199005059 +0800
@@ -1709,7 +1709,7 @@
 {
 	struct drm_i915_gem_create create;
 
-	create.handle = 0;
+	memset(&create, 0, sizeof(create));
 	create.size = size;
 	(void)drmIoctl(fd, DRM_IOCTL_I915_GEM_CREATE, &create);
 
