From 50d576e5269f84da9bd3754500be39119145818a Mon Sep 17 00:00:00 2001
From: Matt Roper <matthew.d.roper@intel.com>
Date: Mon, 17 Jun 2019 14:54:17 -0700
Subject: [PATCH] Add IOTG kernel identifier

---
 drivers/gpu/drm/i915/i915_drv.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index cf328ee..37eb21b 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -1461,6 +1461,9 @@ int i915_driver_load(struct pci_dev *pdev, const struct pci_device_id *ent)
 
 	i915_welcome_messages(dev_priv);
 
+	printk(KERN_INFO "IOTG i915 driver 2020y-02m-25d-14h-43m-07s -0800\n");
+	printk(KERN_INFO "22 patches carried on top of baseline 74ec6e6720\n");
+
 	return 0;
 
 out_cleanup_hw:
-- 
1.9.1

