From 6a508248218cfb68190cbac9892ff96d83d0540f Mon Sep 17 00:00:00 2001
From: Matt Roper <matthew.d.roper@intel.com>
Date: Mon, 17 Jun 2019 14:54:17 -0700
Subject: [PATCH] Add IOTG kernel identifier

---
 drivers/gpu/drm/i915/i915_drv.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index cf328ee..6820a6f 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -1461,6 +1461,9 @@ int i915_driver_load(struct pci_dev *pdev, const struct pci_device_id *ent)
 
 	i915_welcome_messages(dev_priv);
 
+	printk(KERN_INFO "IOTG i915 driver 2020y-05m-20d-16h-45m-37s -0700\n");
+	printk(KERN_INFO "22 patches carried on top of baseline 8abec6c5eb\n");
+
 	return 0;
 
 out_cleanup_hw:
-- 
1.9.1

