From 8a6525ecac47268f63dd995d46412915b9288894 Mon Sep 17 00:00:00 2001
From: Matt Roper <matthew.d.roper@intel.com>
Date: Tue, 3 Jul 2018 14:19:41 -0700
Subject: [PATCH 44/44] Add IOTG kernel identifier

Signed-off-by: Matt Roper <matthew.d.roper@intel.com>
---
 drivers/gpu/drm/i915/i915_drv.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index ad94d7a..b42f8e4 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -1426,6 +1426,9 @@ int i915_driver_load(struct pci_dev *pdev, const struct pci_device_id *ent)
 
 	intel_runtime_pm_put(dev_priv);
 
+	printk(KERN_INFO "IOTG i915 driver 2018y-09m-10d-08h-25m-39s -0700\n");
+	printk(KERN_INFO "43 patches carried on top of PK commit 8730ef32\n");
+
 	return 0;
 
 out_cleanup_hw:
-- 
1.9.1

