From db613ee4a5ea42413e6656e5d439b97ad6de08e0 Mon Sep 17 00:00:00 2001
From: Matt Roper <matthew.d.roper@intel.com>
Date: Mon, 8 Apr 2019 14:22:09 -0700
Subject: [PATCH 46/46] Add IOTG kernel identifier

---
 drivers/gpu/drm/i915/i915_drv.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/gpu/drm/i915/i915_drv.c b/drivers/gpu/drm/i915/i915_drv.c
index e6a07bb..0f5487a 100644
--- a/drivers/gpu/drm/i915/i915_drv.c
+++ b/drivers/gpu/drm/i915/i915_drv.c
@@ -1425,6 +1425,9 @@ int i915_driver_load(struct pci_dev *pdev, const struct pci_device_id *ent)
 
 	intel_runtime_pm_put(dev_priv);
 
+	printk(KERN_INFO "IOTG i915 driver 2019y-04m-08d-14h-22m-09s -0700\n");
+	printk(KERN_INFO "45 patches carried on top of PK commit a0999967\n");
+
 	return 0;
 
 out_cleanup_hw:
-- 
1.9.1

